### ISSUE TITLE
Compatibility BROKEN with PHP 5.6 in 4.0!

### ISSUE THREAD

**Issue by alorbach** [ericmann/wp-session-manager#76]
php --version
PHP 5.6.39-0+deb8u1 (cli) (built: Dec 16 2018 21:56:53)

Since the last update, compatibility to PHP 5.6 is broken. 

Getting multiple errors like these: 
PHP Parse error:  syntax error, unexpected ':', expecting '{' in /xxx/wp-content/plugins/wp-session-manager/includes/deprecated.php on line 15,

This is causing http 500 errors on wordpress sites using this plugin.

**Comment by cyphix333** [comment]
According to the WP plugin page he now has the minimum supported PHP version set to 7.1: https://wordpress.org/plugins/wp-session-manager/

So it seems this change was done on purpose.

**Comment by ericmann** [comment]
PHP 5.6 was end of life last year. PHP 7.0 was end of life last year. PHP 7.1 is the **minimum stable, secure version of PHP available** and the plugin was updated to use it explicitly. The plugin page was updated to reflect the minimum PHP version, and the update included an upgrade notice calling it out.

**You need to upgrade your site's PHP version.**

If you are unable to do so, please remove v4.0.0 and install [version 3.0.4](https://wordpress.org/plugins/wp-session-manager/advanced/) from the "Previous Versions" section at the bottom [of the Advanced page](https://wordpress.org/plugins/wp-session-manager/advanced/) in the plugin repository.

**Comment by alorbach** [comment]
Thats a bad update policy in my opinion. 

Besides the wordpress requirements recommend php 7.2 but they SUPPORT PHP down to PHP 5.2.4!: 
https://wordpress.org/support/article/requirements/

Yes I understand that PHP 5.xxx is outdated, but the descision to update servers with newer version (Which is not possible through packages if you are on older Debian / Ubuntu Versions) is not made by a simple wordpress plugin. 

I have already reversed the update to 3.0.4.

**Comment by ericmann** [comment]
> the descision to update servers with newer version ... is not made by a simple wordpress plugin.

You're right. It's not. But the decision whether or not to support the plugin on older versions of PHP _for free_ is definitely one that's made by the _unpaid_ plugin developer.

Let me be clear. You're perfectly entitled to your opinion here as to whether or not this was a good update or a good call to drop unsupported versions of PHP. At the same time, remember that no one paid for the development of this plugin and no one pays for the ongoing support of this plugin. It's something you and other developers use for free and something I maintain on my own time.

If you want support for older versions of PHP, that's fine. You have two options:

1. Make the changes to the code yourself to support the older versions of PHP and either keep them yourself (it's open source, after all) or submit a pull request back to the project and I'll merge them in.
1. Pay for the support.

I was originally planning to release v4.1.0 today (which would display an admin notice alerting to the outdated version of PHP in use and disable the plugin itself to avoid anything breaking), but will hold off until this weekend if you're interested in option #2 above.

[I've opened a contribution form on my blog](https://ttmm.io/wp-session-manager-support/) and will be sharing the link around this week. **If you use WP Session Manager on a commercial site** and need these additional features, I'd encourage you to contribute. I can't keep supporting certain things for free.

**Comment by alorbach** [comment]
I understand your arguments. 
Thanks for the offer, but I will stay with 3.0.4 for now until we switch to newer server systems with PHP7.
Everything is fine if the plugin disables itself if not supported PHP Version is installed.

**Comment by alorbach** [comment]
I updated to Debian9 which includes PHP7.0 by default. 
I updated to the latest Module, and saw that the requirement is set to 7.1 now.

Is that really necessary? 

BTW, wordpress version.php tells this:
$required_php_version = '5.2.4';

**Comment by ericmann** [comment]
As I said before:
>PHP 5.6 was end of life last year. PHP 7.0 was end of life last year. PHP 7.1 is the **minimum stable, secure version of PHP available** and the plugin was updated to use it explicitly.

Even 7.1 is no longer under active support and is _only_ receiving security patches. It leaves security support later this year, at which point I will update the plugin again to set a minimum version to 7.2.

See: http://php.net/supported-versions.php

WordPress itself (currently) has a minimum requirement of PHP 5.2.4. This version is **also** out of date and insecure. The 5.2 branch stopped receiving support in 2011 - No one should be running production code on PHP 5.2.

> Is that really necessary?

Yes. I will not support older versions of PHP on my own dime and will default to the oldest _supported_ version of PHP available. If anyone needs or wants support for a legacy version (be that 5.6 or even 7.0) they're welcome to contribute to development and I'll work to add support.

**Comment by alorbach** [comment]
The reality is that on production servers systems like Debian are often used and they are very conservative when it comes to using newer versions of anything. 

So right now Debian 9 is stable version and used by most webservers running Debian, and the stable version that comes with Debian 9 is PHP7.0 at the moment. 

This wont change until Debian 10 is released later this year. 
So your plugin won't run on any server running current Debian 9 stable version.

**Comment by cyphix333** [comment]
Whilst I agree not supporting PHP 5.6 is the right call, continually updating the minimum version to the latest _supported_ one is wrong; I believe PHP 7.1 should be the minimum supported version and be this way for awhile; whilst you should have a _recommended_ version, you should continue to support older versions for some time - you will find this is how the majority of software works; _most_ plugins still support 5.6 and WordPress itself 5.2.4 (I believe); you won't find many plugins (of software in general) that continually bump the minimum supported version based off the PHP EOL stance.

**Comment by cyphix333** [comment]
> You are certainly entitled to have an opinion and to make requests. However, you are not entitled to my free time. I've made myself clear in this conversation and elsewhere: I will not support my software on unsupported versions of PHP unless someone is willing to pay for the time required to do so.

The thing is, you keep saying it's going to be extra work to support these "other versions", which I don't think is the case. Unless you're routinely using new cutting edge features added to the new versions or stuff that has been deprecated/removed then there wouldn't be any difference in the different versions, support wise - and I doubt that happens very often.

**Comment by ericmann** [comment]
As I said before:

> If you want support for older versions of PHP, that's fine. You have two options:
>
> 1. Make the changes to the code yourself to support the older versions of PHP and either keep them yourself (it's open source, after all) or submit a pull request back to the project and I'll merge them in.
> 2. Pay for the support.

And also: 

> I will not support older versions of PHP on my own dime and will default to the oldest _supported_ version of PHP available. If anyone needs or wants support for a legacy version (be that 5.6 or even 7.0) they're welcome to contribute to development and I'll work to add support.

The plugin has shipped with [a contribution link](https://paypal.me/eam) for years. I've received $0 in donations. I've both [opened a contribution form on my site](https://ttmm.io/wp-session-manager-support/) and referenced [an existing Patreon page](https://www.patreon.com/user?u=16402577) as well; neither have received any contributions in support of this plugin.

> The thing is, you keep saying it's going to be extra work to support these "other versions", which I don't think is the case.

You're more than welcome to make the change yourself and contribute a patch. But **I will not support my software on unsupported releases of PHP for free**. I'm not sure how more clear I can make that position.

**Comment by glagonikas** [comment]
I don't see the point in this thread. The plugin author has given the following options
1. Pay for his time
2. Use v3.0.4
3. Send a PR yourselves
4. Upgrade your PHP version to a supported one

You keep asking for backward compatibility and not willing to do any of the above, which i find completely unfair, yet you still want to use his **free** plugin??

**Comment by cyphix333** [comment]
> You're more than welcome to make the change yourself and contribute a patch. But I will not support my software on unsupported releases of PHP for free. I'm not sure how more clear I can make that position.

That doesn't really address what I said:

When you plan to only start supporting 7.2, then still supporting 7.1 would require **NO** extra work on your part _unless_ you change the plugin for some reason with new features found in 7.2 or the current version includes code that may not be backward compatible with 7.1, as per: http://php.net/manual/en/migration72.php

When you change the minimum supported version, this is generally because there are new features you wish to use in the newer versions, not just because you want to up the version for no reason other than the current minimum version has reached EOL.

**Comment by ericmann** [comment]
> When you change the minimum supported version, this is generally because there are new features you wish to use in the newer versions

Version 4.0 originally used nullable types, which are only available in 7.1 and above. I removed these not to retain backwards compatibility, but due to a subversion hook in the WordPress plugin repository that erred on their presence. That is a bug that's being fixed, so nullable types _will_ be in a release in the near term.

> not just because you want to up the version for no reason other than the current minimum version has reached EOL.

I will not support EOL versions of PHP for free. Which means I won't test code on those versions. Which means I cannot and will not claim my code runs on those versions. Which is why I bumped the minimum version requirements.

**Comment by cyphix333** [comment]
> Version 4.0 originally used nullable types, which are only available in 7.1 and above. I removed these not to retain backwards compatibility, but due to a subversion hook in the WordPress plugin repository that erred on their presence. That is a bug that's being fixed, so nullable types will be in a release in the near term.

That's fair if you wanted to add new functionality, but as I said, instances such as that are probably few and far between and you're just bumping the minimum supported version because its EOL.

> I will not support EOL versions of PHP for free. Which means I won't test code on those versions. Which means I cannot and will not claim my code runs on those versions. Which is why I bumped the minimum version requirements.

I can understand that to a degree, but perhaps, with all due respect, that you shouldn't be creating free plugins then if you're so concerned without only supporting the most current PHP versions as you don't wish to do the work to check it works on PHP versions that are still widely used - you're effectively forcing people to have to update their PHP version, likely just for your plugin, since I'm sure 99% of others plugins wouldn't have this restriction.

At the end of the day it's your choice, and I know the plugin is free, but you're really inconveniencing your whole user base here.

**Comment by glagonikas** [comment]
@cyphix333 i'm not sure i'm reading your comment correctly. Are you saying Eric is wrong for creating a **FREE** plugin and putting on Github???

**Comment by cyphix333** [comment]
> 
> 
> @cyphix333 i'm not sure i'm reading your comment correctly. Are you saying Eric is wrong for creating a **FREE** plugin and putting on Github???
> 
> So you're saying he should be supporting it for free just because some people decided to use his code?

Of course not - I'm saying he's making it out like creating a free plugin is a hassle and he only wants to support the latest versions; so that's why I said, perhaps that creating a free plugin isn't right for him then.

When you create a plugin, free or not, you have to decide *why* you're creating it and if you pick free OR paid then you should put forth the most effort you can into it whether it's free or paid; I'm sure Eric does put a lot of effort into developing it as it's a good plugin, but because it appears to be somewhat of a hassle for him to support earlier versions of PHP (even on the 7 branch) perhaps he should have gone down the paid route instead as he doesn't seem to be committed to providing a free version unless you happen to meet certain guidelines.

Since it is a free plugin, Eric doesn't owe anyone anything, again it's his choice, but since he decided to go the free route, then he should be fully committed to providing a well supported free version; it's not a niche situation where people are expecting support for things that will only help a few people, it's about realistically supporting your userbase.

An example would be, for many years many developers would still have to support IE6, as people were still using it, even though it was very old and better browsers were out there, but it stayed supported because of the % of people still using it - but over time the % dropped and IE6 became like **10 years old** and people started dropping support.

With that in mind, it's completely reasonable to not support 5.6 (even though most plugins still do), because it is so old, but continually bumping the minimum supported version isn't the standard way to do things.

**Comment by ericmann** [comment]
> When you create a plugin, free or not, you have to decide _why_ you're creating it 

Simple. [I created this as a proof-of-concept feature plugin for a core WordPress feature I wanted added](https://ttmm.io/tech/introducing-wp_session/). I was told at the time that any such addition _must_ start as a plugin and downloads/adoption would be used to gate merging into core. That's why it exists and why it was listed on WordPress.org.

> I'm sure Eric does put a lot of effort into developing it as it's a good plugin

Thanks.

> because it appears to be somewhat of a hassle for him to support earlier versions of PHP (even on the 7 branch) perhaps he should have gone down the paid route instead as he doesn't seem to be committed to providing a free version unless you happen to meet certain guidelines.

I do provide a free version. It's right here. Use it if you want. Or don't. I really don't care. I wrote this as a PoC and, unbeknownst to me, it was bundled into other plugins: WooCommerce, Easy Digital Downloads, Ninja Forms, etc. I _continue_ to maintain the plugin so I can patch bugs as they come up and add features that support various use cases I'm interested in - like encrypting sessions stored at rest or [experimenting with a middleware-approach to native session management](https://ttmm.io/tech/introducing-sessionz/).

> Since it is a free plugin, Eric doesn't owe anyone anything

You are 100% right with this.

> since he decided to go the free route, then he should be fully committed to providing a well supported free version

Nope. This is a direct contradiction of your previous point. I owe you nothing. I don't owe you additional features or support. You (and others) **have paid and contributed nothing** to the support or development of this plugin.

> With that in mind, it's completely reasonable to not support 5.6 (even though most plugins still do), because it is so old, but continually bumping the minimum supported version isn't the standard way to do things.

Again, you're entitled to your opinion, but I'm killing this conversation now. Unless someone is willing to pay for my time, I will not support my tools on unsupported versions of the core language they use. Full stop.

There seems to be nothing I or anyone else can say to help you understand this point.
